libursa/QueryGraph.cpp

```diff
@@ -29,40 +32,116 @@ QueryGraph QueryGraph::dual() const {
 
+    // Moves the provided value into internal state.
+    void set(T &&state) { state_ = std::move(state); }
+
+    // Gets a vector of already processed predecessors.
     const std::vector<NodeId> &ready_predecessors() const {
         return ready_predecessors_;
     }
 
-    void set(QueryResult &&state) { state_ = std::move(state); }
-
+    // Used for counting incoming edges into the graphs node.
     void add_predecessor() { total_predecessors_++; }
 
+    // Called when a predecessor was processed and becomes ready.
     void add_ready_predecessor(NodeId ready) {
         ready_predecessors_.push_back(ready);
     }
 
+    // Node is assumed ready when all of its predecessors were processed.
     bool ready() const {
         return ready_predecessors_.size() >= total_predecessors_;
     }
 };
 
-QueryResult masked_or(const std::vector<const QueryResult *> &to_or,
+// This class can be used to traverse QueryGraph DAG in topological order.
+// It's guaranteed that during processing of every node, all it's parents
+// were already processed.
+template <typename T, T TM()>
+class InorderGraphVisitor {
+    std::vector<NodeId> ready_;
+    const std::vector<QueryGraphNode> *nodes_;
+    std::vector<NodeState<T, TM>> state_;
+
+   public:
+    // Creates a new instance of InorderGraphVisitor. `ready` vector should
+    // contain verticles with no incoming edges (aka graph sources).
+    InorderGraphVisitor(std::vector<NodeId> ready,
+                        const std::vector<QueryGraphNode> *nodes)
+        : ready_(std::move(ready)), nodes_(nodes), state_() {
+        state_.reserve(nodes_->size());
+        for (size_t i = 0; i < nodes_->size(); i++) {
+            state_.push_back(NodeState<T, TM>());
+        }
+        for (size_t ndx = 0; ndx < nodes_->size(); ndx++) {
+            for (NodeId target : (*nodes)[ndx].edges()) {
+                state_.at(target.get()).add_predecessor();
+            }
+        }
+    }
+
+    // True if traversal was finished and visit queue is empty.
+    bool empty() const { return ready_.empty(); }
+
+    // Returns id of next graph's node to process.
+    NodeId next() {
+        NodeId nextid = ready_.back();
+        ready_.pop_back();
+        for (const auto &succ : (*nodes_)[nextid.get()].edges()) {
+            state_[succ.get()].add_ready_predecessor(nextid);
+            if (state_[succ.get()].ready()) {
+                ready_.push_back(succ);
+            }
+        }
+        return nextid;
+    }
+
+    // Returns a vector of states of all predecessors for a given node.
+    std::vector<const T *> predecessor_states(NodeId id) {
+        std::vector<const T *> states;
+        const NodeState<T, TM> &state = state_[id.get()];
+        states.reserve(state.ready_predecessors().size());
+        for (const auto &pred : state.ready_predecessors()) {
+            states.push_back(&state_[pred.get()].get());
+        }
+        return std::move(states);
```

Comment:
Seems unnecessary because of NRVO

Reply:
more std::moves = the program goes faster, it is known

Reply:
This is just a vector of pointers so it doesn't matter much in this case, but do we *really* want to trust the compiler to elide the copy?

Reply:
For reference:

https://stackoverflow.com/questions/14856344/when-should-stdmove-be-used-on-a-function-return-value

Gosh.